### PR TITLE
DW-183 / DW-218

### PR DESCRIPTION
**Pull request by Kucki99**

### CODE REVIEW

**Comment by Kucki99** [comment]
Missing: 
- unit/selenium tests
- change password (for logged in users)

File: src/main/java/de/digiwill/controller/ProfileController.java [hunk 1]
```diff
@@ -1,27 +1,77 @@
...
+        return new ModelAndView(response.isSuccess() ? "redirect:/profile" : "redirect:/editProfile");
+    }
+
+    private void addFlashAttributesToRedirectAttributes(RedirectAttributes redirectAttrs, ValidationResponse response) {
+        redirectAttrs.addFlashAttribute("edit", !response.isSuccess());
+        redirectAttrs.addFlashAttribute("hasToast", !response.isSuccess());
+        redirectAttrs.addFlashAttribute("responseText", response.getFailureMessage());
+    }
+
+    private void addDataToProfilePageModel(Principal principal, Model model, boolean edit) {
```
Comment (MoellJ) [inline src/main/java/de/digiwill/controller/ProfileController.java]: This method seems like it should be in the profileService

File: src/test/resources/de.digiwill/editProfile.feature [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    And I clear field with id "surNameInput"
+    And I enter "<surName>" into field with id "surNameInput"
+    And I clear field with id "birthday"
+    And I enter "<birthday>" into field with id "birthday"
+    And I enter "<streetAddress>" into field with id "streetAddressInput"
+    And I enter "<zipCode>" into field with id "zipCodeInput"
+    And I enter "<city>" into field with id "cityInput"
+    And I enter "<country>" into field with id "countryInput"
+    And I click on element with id "saveProfileButton"
+    Then I'm on page with title "DigiWill - Your Profile"
```
Comment (MoellJ) [inline src/test/resources/de.digiwill/editProfile.feature]: Shouldn't we check, that the information actually got updated?

File: src/test/resources/de.digiwill/editProfile.feature [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      | Tester    | McTest  | 04201990 | Test-Street 41 | 12345   | Test-City | Germany |
+
+  Scenario Outline: 02 - Change password successful
+    When I click on element with id "userButtonHeader"
+    And I click on element with id "editProfileButton"
+    And I enter "<oldPassword>" into field with id "oldPasswordInput"
+    And I enter "<newPassword>" into field with id "newPasswordInput"
+    And I enter "<repNewPassword>" into field with id "confirmNewPasswordInput"
+    And I click on element with id "changePasswordButton"
+    Then I'm on page with title "DigiWill - Your Profile"
```
Comment (MoellJ) [inline src/test/resources/de.digiwill/editProfile.feature]: Shouldn't we check, that the user can now log in with the new password?

File: src/main/java/de/digiwill/controller/ProfileController.java [hunk 4]
```diff
@@ -1,27 +1,77 @@
...
+    }
+
+    @GetMapping("/editProfile")
+    public String editProfile(Principal principal, Model model, RedirectAttributes redirectAttrs) {
+        addDataToProfilePageModel(principal, model, true);
+
+        return "profile";
+    }
+
+    @RequestMapping(value = "/changePassword", method = RequestMethod.POST,
```
Comment (MoellJ) [inline src/main/java/de/digiwill/controller/ProfileController.java]: Replace with     @PostMapping(value = "/changePassword", consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)

File: src/main/java/de/digiwill/controller/ProfileController.java [hunk 5]
```diff
@@ -1,27 +1,77 @@
...
+            consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)
+    public ModelAndView changePassword(@RequestBody MultiValueMap<String, String> formData, Model model, Principal principal, RedirectAttributes redirectAttrs) {
+        ValidationResponse response = profileService.changePassword(formData, principal.getName());
+        response.adjustModel(model);
+        addFlashAttributesToRedirectAttributes(redirectAttrs, response);
+        return new ModelAndView(response.isSuccess() ? "redirect:/profile" : "redirect:/editProfile");
+    }
+
+    @RequestMapping(value = "/editProfile", method = RequestMethod.POST,
```
Comment (MoellJ) [inline src/main/java/de/digiwill/controller/ProfileController.java]: Replace with     @PostMapping(value = "/editProfile",  consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE)

File: src/test/resources/de.digiwill/editProfile.feature (hunk 2 above)
Comment (Kucki99) [reply src/test/resources/de.digiwill/editProfile.feature]: This makes sense. Do you have a suggestion on how we can test this?

File: src/main/java/de/digiwill/service/ProfileService.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        editProfileValidators.add(new NonEmptyStringValidator("firstNameInput", ValidationResponse.NO_FIRST_NAME));
+        editProfileValidators.add(new NonEmptyStringValidator("surNameInput", ValidationResponse.NO_SURNAME));
+        editProfileValidators.add(new BirthdayValidator());
+
+        changePasswordValidators = new ArrayList<>();
+        changePasswordValidators.add(new PasswordRequirementValidator());
+        changePasswordValidators.add(new PasswordMatchValidator());
+    }
+
+    public UserHandle getUserHandleByEmail(String email) {
```
Comment (MoellJ) [inline src/main/java/de/digiwill/service/ProfileService.java]: Pointless

**Pull request merged**